pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/ClusterChangeMediator.java

```diff
@@ -18,142 +18,136 @@
+              } else {
+                long lastProcessTime = _lastProcessTimeMap.get(changeType);
+                if (currentTime - lastProcessTime > PROACTIVE_CHANGE_CHECK_INTERVAL_MS) {
+                  LOGGER.info("Proactive check {} change", changeType);
+                  brokerMetrics.addMeteredGlobalValue(BrokerMeter.PROACTIVE_CLUSTER_CHANGE_CHECK, 1L);
+                  processClusterChange(changeType, changeHandler);
+                }
               }
             }
 
-            if (instanceConfigUpdated) {
-              try {
-                _helixExternalViewBasedRouting.processInstanceConfigChange();
-              } catch (Exception e) {
-                LOGGER.warn("Caught exception while processing instance config", e);
-              }
-            }
+            // Sleep 1 second after each round of checks
+            Thread.sleep(1000L);
           } catch (InterruptedException e) {
-            LOGGER.warn("Was interrupted while waiting for a cluster change", e);
+            LOGGER.warn("Cluster change handling thread is interrupted, stopping the thread");
             break;
+          } catch (Exception e) {
+            LOGGER.error("Caught exception while handling changes", e);
           }
         }
-
-        LOGGER.warn("Stopping deferred cluster state update thread");
-        _deferredClusterUpdater = null;
       }
     };
+    _clusterChangeHandlingThread.start();
+  }
 
-    _deferredClusterUpdater.start();
+  private void processClusterChange(ChangeType changeType, ClusterChangeHandler changeHandler) {
+    long startTime = System.currentTimeMillis();
+    LOGGER.info("Start processing {} change", changeType);
+    changeHandler.processClusterChange();
+    long endTime = System.currentTimeMillis();
+    LOGGER.info("Finish processing {} change in {}ms", changeType, endTime - startTime);
+    _lastProcessTimeMap.put(changeType, endTime);
   }
 
   @Override
   public void onExternalViewChange(List<ExternalView> externalViewList, NotificationContext changeContext) {
-    // If the deferred update thread is alive, defer the update
-    if (_deferredClusterUpdater != null && _deferredClusterUpdater.isAlive()) {
-      try {
-        _clusterChangeQueue.put(new ImmutablePair<>(UpdateType.EXTERNAL_VIEW, System.currentTimeMillis()));
-      } catch (InterruptedException e) {
-        LOGGER.warn("Was interrupted while trying to add external view change to queue", e);
-      }
-    } else {
-      LOGGER.warn(
-          "Deferred cluster updater thread is null or stopped, not deferring external view routing table rebuild");
-      _helixExternalViewBasedRouting.processExternalViewChange();
-      _tableQueryQuotaManager.processQueryQuotaChange();
-    }
+    Preconditions.checkState(externalViewList.isEmpty(), "Helix pre-fetch should be disabled");
+    Preconditions.checkState(_clusterChangeHandlingThread.isAlive(), "Cluster change handling thread is not alive");
```

Comment:
This can happen during a race condition where the thread takes longer to start up? maybe you should add a condition to wait until thread is alive before adding these listeners

Reply:
> A thread is alive if it has been started and has not yet died.